homeassistant/components/advantage_air/climate.py

```diff
@@ -82,39 +83,116 @@ async def async_setup_entry(
 class AdvantageAirAC(AdvantageAirAcEntity, ClimateEntity):
     """AdvantageAir AC unit."""
 
+    _attr_fan_modes = [FAN_LOW, FAN_MEDIUM, FAN_HIGH]
     _attr_temperature_unit = TEMP_CELSIUS
     _attr_target_temperature_step = PRECISION_WHOLE
     _attr_max_temp = 32
     _attr_min_temp = 16
-    _attr_fan_modes = [FAN_AUTO, FAN_LOW, FAN_MEDIUM, FAN_HIGH]
-    _attr_hvac_modes = AC_HVAC_MODES
-    _attr_supported_features = (
-        ClimateEntityFeature.TARGET_TEMPERATURE | ClimateEntityFeature.FAN_MODE
-    )
 
     def __init__(self, instance, ac_key):
         """Initialize an AdvantageAir AC unit."""
         super().__init__(instance, ac_key)
-        if self._ac.get("myAutoModeEnabled"):
-            self._attr_hvac_modes = AC_HVAC_MODES + [HVACMode.AUTO]
+        self._attr_unique_id = f'{self.coordinator.data["system"]["rid"]}-{ac_key}'
+
+        self._attr_preset_modes = [ADVANTAGE_AIR_MYZONE]
+        self._attr_supported_features = ClimateEntityFeature.FAN_MODE
+
+        # Add "MyTemp" preset if available
+        if ADVANTAGE_AIR_MYTEMP_ENABLED in self._ac:
+            self._attr_preset_modes += [ADVANTAGE_AIR_MYTEMP]
+            self._attr_supported_features |= ClimateEntityFeature.PRESET_MODE
+
+        # Add "MyAuto" preset if available
+        if ADVANTAGE_AIR_MYAUTO_ENABLED in self._ac:
+            self._attr_preset_modes += [ADVANTAGE_AIR_MYAUTO]
+            self._attr_supported_features |= ClimateEntityFeature.PRESET_MODE
 
     @property
-    def target_temperature(self):
+    def target_temperature(self) -> float | None:
         """Return the current target temperature."""
         return self._ac["setTemp"]
 
     @property
-    def hvac_mode(self):
+    def hvac_mode(self) -> HVACMode:
         """Return the current HVAC modes."""
         if self._ac["state"] == ADVANTAGE_AIR_STATE_ON:
-            return ADVANTAGE_AIR_HVAC_MODES.get(self._ac["mode"])
+            return ADVANTAGE_AIR_HVAC_MODES[self._ac["mode"]]
         return HVACMode.OFF
 
     @property
-    def fan_mode(self):
+    def hvac_modes(self) -> list[HVACMode]:
+        """Return the available HVAC modes."""
```

Comment:
Can you set this as `self._attr_hvac_modes`?

Reply:
No because the available HVAC modes change depending on which preset is selected.

Reply:
We aren't allowed to change the hvac modes after the start event. If the hvac mode can't be set in the current state it should raise an exception instead.  See the silver requirements in the devs docs for which one to raise

Reply:
So what I've implemented works perfectly in the platform and the UI, so is this an unwritten rule?

Would the acceptable implementation be to show every HVAC mode but just throw errors if its not valid for the current state, despite the UI showing all options?

Reply:
> So what I've implemented works perfectly in the platform and the UI, so is this an unwritten rule?

All the states are sent to the voice assistants / forwarding integrations at the started event.  If you change the model out from under them they won't be able to access the states that were not visible to them at startup.

> 
> Would the acceptable implementation be to show every HVAC mode but just throw errors if its not valid for the current state, despite the UI showing all options?

Assuming you mean every HVAC mode the device can support, then Yes.

Reply:
Is there any way to differentiate what is sent to forwarding integrations at startup and what is shown in the WebUI, because obviously this is going to degrade the user experience showing options that don't work.

It is not common to change the operating mode, so I was willing to accept that voice assistants may not work when the preset has changed.

Reply:
> Is there any way to differentiate what is sent to forwarding integrations at startup and what is shown in the WebUI, because obviously this is going to degrade the user experience showing options that don't work.

Each of big ones would have to be rewritten to be able to support that.  At least for homekit, it doesn't like changing the supported features of the entity since it has to resync with iCloud.  I believe alexa and google assistant need similar resyncing when the features the device supports changes.   From an architecture perceptive there isn't much we can change here since most of them only build the devices/structures once for performance reasons, and that seems unlikely to change in the future.

> It is not common to change the operating mode, so I was willing to accept that voice assistants may not work when the preset has changed.

Its also find to require a restart to change it if you want to set the modes once.

Reply:
I understand the limitation from Voice Assistants, so can I instead just accept as a design decision that voice assistants may not work when the preset changes, and it requires a restart (or sync) for voice assistants to understand the new mode?

This product has its own direct integration with voice assistants, so the goal here is to add full functionality to **Home Assistant**, not to the voice assistants.

Reply:
@bdraco is this what you mean by raising ValueError?
https://github.com/Bre77/core/blob/aa-autoselect2/homeassistant/components/advantage_air/climate.py#L166-L175